Event.php

```diff
@@ -63,40 +63,97 @@ public function render(OutputPage &$out) {
         );
 
-        if($e->{'place'}) {
-          $event['location'] = array(
-            '@type' => 'Place',
-            'name'  => $e->{'place'},
-            'address' => array(
-              'streetAddress'   => $this->getMetadataField($e, 'streetAddress', 'streetaddress'),
-              'addressLocality' => $this->getMetadataField($e, 'addressLocality', 'locality'),
-              'postalCode'      => $this->getMetadataField($e, 'postalCode', 'postalcode'),
-              'addressRegion'   => $this->getMetadataField($e, 'addressRegion', 'region'),
-              'addressCountry'  => $this->getMetadataField($e, 'addressCountry', 'country'),
-            ),
+        switch ($e->{'attendancemode'}) {
+          case "MixedEventAttendanceMode":
+            if($e->{'place'}) {
+              $event['location'] = array(
+                $v['online'] = array(
+                  '@type' => "VirtualLocation",
+                  'url'   => $this->getRawURL($e->{'virtualvendorurl'}),
+                ),
+                $p['offline'] = array(
+                  '@type' => 'Place',
+                  'name'  => $e->{'place'},
+                  'address' => array(
+                    'streetAddress'   => $this->getMetadataField($e, 'streetAddress', 'streetaddress'),
+                    'addressLocality' => $this->getMetadataField($e, 'addressLocality', 'locality'),
+                    'postalCode'      => $this->getMetadataField($e, 'postalCode', 'postalcode'),
+                    'addressRegion'   => $this->getMetadataField($e, 'addressRegion', 'region'),
+                    'addressCountry'  => $this->getMetadataField($e, 'addressCountry', 'country'),
+                  ),
+                ),
+              );                
+            };
+            break;
+          case "OfflineEventAttendanceMode":
+            if($e->{'place'}) {
+              $event['location'] = array(
+                '@type' => 'Place',
+                'name'  => $e->{'place'},
+                'address' => array(
+                  'streetAddress'   => $this->getMetadataField($e, 'streetAddress', 'streetaddress'),
+                  'addressLocality' => $this->getMetadataField($e, 'addressLocality', 'locality'),
+                  'postalCode'      => $this->getMetadataField($e, 'postalCode', 'postalcode'),
+                  'addressRegion'   => $this->getMetadataField($e, 'addressRegion', 'region'),
+                  'addressCountry'  => $this->getMetadataField($e, 'addressCountry', 'country'),
+                ),
+              );
+            };
+            break;
+          case "OnlineEventAttendanceMode":
+            if($e->{'virtualvendorurl'}) {
+              $event['location'] = array(
+                '@type' => "VirtualLocation",
+                'url'   => $this->getRawURL($e->{'virtualvendorurl'}),
+              );
+            };
+            break;
+          default:
```

Comment:
I'm not well versed with PHP so I kept the default case, not knowing the best way to exit the select statement.

Another area of concern is within the MixedEventAttendanceMode case where there is now two arrays -- nested under locaton  -- not sure if okay to declare them on the fly.

$v['online'] = array(...
and 
$p['offline'] = array(...

The JSON-LD output looks okay and the SDTT tests okay on all three cases. I wanted to try to get this done before Google deprecates the SDTT tool, the Rich Snippets tool has no debugging at present.

As for some of the other optional properties I added, like sameAs etc., SDTT does not seem to care if they are empty.

My wiki is version 133.1 

GoogleRichCards is a great tool for MediaWiki, thanks so much for creating it.

Reply:
> I'm not well versed with PHP so I kept the default case, not knowing the best way to exit the select statement.
> Edited: (Of the three conditions, none of them seemed to be a candidate for default)

I meant it requires to have some default instead of reporting an error in that place to avoid the case users can update their extension and receive exactly the same behaviour they had. It's about backward compatibility with previous version of the extension ;)

About the candidate for attendance mode, we know for sure most of the meetings are meant to be offline i.e. with actual presence but some of them allowed to attend online with online translations, chats, whatsoever. So it looks like the default and the least invasive way here is to use `MixedEventAttendanceMode` as a default since it reflects most of the cases. At the same time it's required to be reflected in README to highlight that.

I assume it's better than getting an error right on the page without any other notice since mediawiki will do exactly that on sending an error. So if you have any other view of that problem - please feel free to describe your point, it's not a requirement here and I'm open to discuss it to provide the better experience for users.

> Another area of concern is within the MixedEventAttendanceMode case where there is now two arrays -- nested under locaton -- not sure if okay to declare them on the fly.
>
> $v['online'] = array(...
> and
> $p['offline'] = array(...
PHP is all about "on the fly" since it's an interpreter. It translates all your code in linear manner line-by-line(actually not, but it's for better illustration). So it's OK for PHP but wouldn't be OK for any statically typed compiler language since memory allocations have their own cost(actually it depends on exact case and for better illustration again :)

> GoogleRichCards is a great tool for MediaWiki, thanks so much for creating it.
I really glad it become helpful and thank you for contribution :)

Reply:
how about setting a case where we make sure that it fails the SDTT?

~~~~
default:
/* Expected cases failed, give user debug information via Structured Data Testing Tool  */
 $event['location'] = array(
                '@type' => "VirtualLocation",
                'url'   => "none",
              );
~~~~

Reply:
I'm not quite sure SDTT is the right to tool to cover negative test case, it should be the aim for unit tests (which are not present here ¯\_(ツ)_/¯ ) while adding such thing to use with SDTT could make some confuse on how to use it and why it's present.

One more thing about it - it's up to SDTT to fail on wrong input, why third-party software should test it?

Reply:
The only other thing I could think of is to write to a MediaWiki log file, but I would have no idea where to start on that.

Just seems like a graceful way to fail the condition.

1. The Wiki doesn't crash / display unwanted message
2. The JSON-LD is incorrect -- not a platform error, just incorrect output.
3. The case will only fail if the enumeration is incorrect 

Additionally, when one of the "if" conditions fails, there is no debugging information.  Everything would have to be wrapped in a 'try" statement?

Reply:
The thing is logging that staff won't change it will be "silent" failure since on the step of upgrade & configuration there's not way to determine it will not work.

To get an error/error entry in log/trigger the extension code in any other way requires some actions after installation such as manually walking through the all the pages using the extension templates and looking into logs. Looks pretty unhandy actions with not strict results (since template configuration could differ and contain different fields from page to page).

Anyway of you think there's no option to set default probably we could describe upgrade notes and explicitly set it backward incompatible - that's the option too.

Reply:
In many cases, JSON-LD output will be wrapped in Result=OK or some other coded error etc. 
This program is expecting specific enumeration, with camel-case for conversion.

I suppose we could create a coding system like above, but simply reporting the exception in the exported JSON-LD seems appropriate.

using "url" : none would simply trip the SDTT to throw an error to help with debugging and then we could describe the error in the description property.  I think that most people will find a way to read the JSON-LD to ensure that it is correct.

~~~~
default:
/* Expected cases failed, give user debug information via Structured Data Testing Tool  */
 $event['location'] = array(
                '@type' => "VirtualLocation",
                'url'   => "none",
                'description => 'Error: attendancemode expected one of the following enumerations: MixedEventAttendanceMode OfflineEventAttendanceMode OnlineEventAttendanceMode,'
              );

~~~~

Reply:
Sorry for long reply.

Here's mainly the thing: as a wiki owner/administrator do you want to check JSON or page source code at all?
I suppose it's a wrong way here, since it requires some knowledge and some time.

That's mainly why I'd prefer the original behaviour to preserve as a default and just add a way to adjust it if needed :)